---
title: "<center> OPLA POLA RNAseq analysis <center>"
author: "<center> Felix Westcott <center><br>"
date: "<center> _`r Sys.Date()`_ <center>"
---

```{r setup, include=FALSE}


library(tidyverse)
library(edgeR)
library(limma)
library(FactoMineR)
library(factoextra)
library(cowplot)
library(gt)
library(DT)
library(ggplot2)
library(EnhancedVolcano)
library(ggpubr)
library(gplots)
library(venn)
library(patchwork)
library(GenomeInfoDbData)
library(org.Hs.eg.db)
library(clusterProfiler)
library(pathview)
library(ggsci)
library(ggprism)
library(Glimma)

```

``` {R Loading data}
# Load data
Raw_counts <- read.delim ('Counts.txt')
samplemetadata <-read.csv ('Sample_info.csv')

group_media <- factor(samplemetadata$media)
group_exp <- factor(samplemetadata$exp)
group_conc <- factor(samplemetadata$concentration)

# Turn raw count data into matrix with only relevant parts
geneID <- Raw_counts$gene_name
Raw_counts <- as.matrix(Raw_counts[,2:26])
rownames(Raw_counts) <- geneID

# Create a DGEList from raw count matrix
myDGEList <- DGEList(Raw_counts, group = group_media)

log2.cpm <- cpm(myDGEList, log=TRUE)


log2.cpm.df <- as_tibble(log2.cpm, rownames = "geneID")
log2.cpm.df.pivot <- pivot_longer(log2.cpm.df,
                                  cols = EC1C:EC5P8, 
                                  names_to = "samples",
                                  values_to = "expression")

ggplot(log2.cpm.df.pivot) +
  aes(x=samples, y=expression, fill=samples) +
  geom_violin(trim = FALSE, show.legend = FALSE) +
  stat_summary(fun = "median", 
               geom = "point", 
               shape = 95, 
               size = 10, 
               color = "black", 
               show.legend = FALSE) +
  labs(y="log2 expression", x = "sample",
       title="Log2 Counts per Million (CPM)",
       subtitle="unfiltered, non-normalized") +
  theme_bw() +
   theme(axis.text.x = element_text(size = 7))
```

``` {R Filter and normalise data}

# Filter data. NB this uses cpm but this will get calculated properly again later
cpm <- cpm(myDGEList, log=FALSE)
lcpm <- cpm(myDGEList, log=TRUE)
keep <- filterByExpr(cpm, group = group_media)
myDGEList.filtered <- myDGEList[keep,,keep.lib.sizes=FALSE]

# Normalize filtered data using TMM method
myDGEList.filtered.norm <- calcNormFactors(myDGEList.filtered, method = "TMM")

# Express filtered and normalised data in log2(cpm)
log2.cpm.filtered.norm <- cpm(myDGEList.filtered.norm, log=TRUE)

# Also have it as a dataframe in case it comes in handy
log2.cpm.filtered.norm.df <- as_tibble(log2.cpm.filtered.norm, rownames = "geneID")

log2.cpm.filtered.norm.df.pivot <- pivot_longer(log2.cpm.filtered.norm.df, 
                                                cols = EC1C:EC5P8, 
                                                names_to = "samples", 
                                                values_to = "expression")

ggplot(log2.cpm.filtered.norm.df.pivot) +
  aes(x=samples, y=expression, fill=samples) +
  geom_violin(trim = FALSE, show.legend = FALSE) +
  stat_summary(fun = "median", 
               geom = "point", 
               shape = 95, 
               size = 10, 
               color = "black", 
               show.legend = FALSE) +
  labs(y="log2 expression", x = "sample",
       title="Log2 Counts per Million (CPM)",
       subtitle="filtered, TMM normalized") +
  theme_bw() +
   theme(axis.text.x = element_text(size = 7))

```

``` {R Gene Expression Data}

# Creating a log2cpm gene expression df with averages for each condition
Avg.gene_expression <- log2.cpm.filtered.norm.df %>% 
  mutate(Control = (EC1C + EC2C + EC3C + EC4C + EC5C)/5,
         OPLA200 = (EC1O2 + EC2O2 + EC3O2 + EC4O2 + EC5O2)/5,
         OPLA800 = (EC1O8 + EC2O8 + EC3O8 + EC4O8 + EC5O8)/5,
         POLA200 = (EC1P2 + EC2P2 + EC3P2 + EC4P2 + EC5P2)/5,
         POLA800 = (EC1P8 + EC2P8 + EC3P8 + EC4P8 + EC5P8)/5) %>% 
    mutate_if(is.numeric, round, 2) %>%
  dplyr::select(geneID, Control, OPLA200, OPLA800,POLA200,POLA800)

datatable(Avg.gene_expression, 
          extensions = c('KeyTable', "FixedHeader"), 
          filter = 'top',
          caption = 'Table 1: log2 CPM of filtered and normalised genes in Control, OPLA and POLA',
          options = list(keys = TRUE, 
                         searchHighlight = TRUE, 
                         pageLength = 10, 
                         lengthMenu = c("10", "25", "50", "100")))

```

```{R, Really cool PCA browser pop out}

#myDGEList.filtered.norm$samples$round <- group_exp
#myDGEList.filtered.norm$samples$conc <- group_conc

#glMDSPlot(lcpm, labels=paste(group_media, group_exp, group_exp, sep="_"), groups=myDGEList.filtered.norm$samples[,c(1,4,5)], launch=TRUE)

```


``` {R PCA analysis and scree plot}
#This is exactly the same as before, I just added the summary print out

pca.res <- prcomp(t(log2.cpm.filtered.norm), scale.=F, retx=T)

summary(pca.res)

fviz_eig(pca.res, addlabels = TRUE, ylim = c(0,100))

```

``` {R individual PCA plot}
#This is exactly the same as before I think

fviz_pca_ind(pca.res, geom.ind = "point", pointshape = 21, 
             pointsize = 2, 
             fill.ind = group_exp,
             palette = "npg",
             col.ind = "black", 
             addEllipses = TRUE,
             label = "var",
             col.var = "black",
             repel = TRUE,
             legend.title = "Group", title = "Principle Components Analysis") +
  theme_prism(base_size = 14)

#ggsave("C:/Users/fwestcott/OneDrive - Nexus365/Desktop/Outputs/Transfer-report/Figures/PCA.png")

```


``` {R PCA 'small mulitples' plots}

# This is all new so I'll explain what I'm doing.

# Take results of first four PCAs and turn into tibble with sample names and fat types as added rows

pca.res.df <- pca.res$x[,1:4] %>%
  as_tibble() %>%
  add_column(sample = samplemetadata$sample_name,
             group = group_exp)
  
#Pivot the table to make it tidy and easier to facet graph

pca.pivot <- pivot_longer(pca.res.df, 
                          cols = PC1:PC4, 
                          names_to = "PC", 
                          values_to = "loadings")

#Graph - this graph illustrates if there are any obvious confounding variables among your samples (ie batch effects etc.)

ggplot(pca.pivot) +
  aes(x=sample, y=loadings, fill=group) + 
  geom_bar(stat="identity") +
  facet_wrap(~PC) +
  labs(title="PCA 'small multiples' plot") +
  theme_bw() +
  coord_flip() +
theme(axis.text.x = element_text(color = "black", size = 8, angle = 0, hjust = .5, vjust = .5, face = "plain"),
      axis.text.y = element_text(color = "black", size = 8, angle = 0, hjust = .5, vjust = .5, face = "plain"),
      axis.title.x = element_text(color = "black", size = 12, angle = 0, hjust = .5, vjust = .5, face = "plain"),
      axis.title.y = element_text(color = "black", size = 12, angle = 90, hjust = .5, vjust = .5, face = "plain"))

```

``` {R Differential Gene Expression Analysis!}

# Pretty much all the same as before I just tidied this up quite a bit and put it all in one place for you

# Design matrix - exactly the same as before
design <- model.matrix(~0 + group)
colnames(design) <- levels(group)

# Contrast matrix - slight changes to before just to tidy up
contrast.matrix1 <- makeContrasts(OPLAvsControl = OPLA-Control, levels=design)
contrast.matrix2 <- makeContrasts(POLAvsControl = POLA-Control, levels=design)
contrast.matrix3 <- makeContrasts(POLAvsOPLA = POLA-OPLA, levels=design)

# Estimate common dispersion, trended dispersions and tagwise dispersions from filtered and normalized raw data (ie not cpm or log)
y <- estimateDisp(myDGEList.filtered.norm, design, robust = TRUE)
fit <- glmFit(y, design)

# Likelihood ratio test OPLA vs Control. 
res1 <- glmLRT(fit, contrast = contrast.matrix1)
results1 <- topTags (res1, sort.by = NULL, n=Inf, adjust.method = "BH")
toptable1 <- results1$table %>%
  as_tibble(rownames = "geneID")

# Likelihood ratio test POLA vs Control.
res2 <- glmLRT(fit, contrast = contrast.matrix2)
results2 <- topTags (res2, sort.by = NULL, n=Inf, adjust.method = "BH")
toptable2 <- results2$table %>%
  as_tibble(rownames = "geneID")

# Likelihood ratio test POLA vs OPLA. 
res3 <- glmLRT(fit, contrast = contrast.matrix3)
results3 <- topTags (res3, sort.by = NULL, n=Inf, adjust.method = "BH")
toptable3 <- results3$table %>%
  as_tibble(rownames = "geneID")

# NB: FDR column is adjusted p-value using BH correction for multiple comparisons

DGEResults <- cbind(toptable1$geneID, toptable1$logFC, toptable2$logFC, toptable3$logFC, toptable1$FDR, toptable2$FDR, toptable3$FDR) %>%
  as.data.frame() %>%
  rename(geneID = V1, "OPLA vs Control LogFC" = V2, "POLA vs Control logFC" = V3, "POLA vs OPLA logFC" = V4, "OPLA vs Control p.adj" = V5, "POLA vs Control p.adj" = V6, "POLA vs OPLA p.adj" = V7)

#gt(DGEResults)
```

``` {R Volcano plots!!}

# Almost exactly the same just all in one place and made to look better 

EnhancedVolcano(toptable1,
    lab = toptable1$geneID,
    x = 'logFC',
    y = 'FDR',
    legendPosition = "right",
    legendLabSize = 12,
    legendIconSize = 3,
    labSize = 3,
    subtitle = "OPLA vs Control",
    pCutoff = 0.01, 
    FCcutoff = 0.2)


EnhancedVolcano(toptable2,
    lab = toptable2$geneID,
    x = 'logFC',
    y = 'FDR',
    legendPosition = "right",
    legendLabSize = 12,
    legendIconSize = 3,
    labSize = 3,
    subtitle = "POLA vs Control",
    pCutoff = 0.01, 
    FCcutoff = 0.2)


EnhancedVolcano(toptable3,
    lab = toptable3$geneID,
    x = 'logFC',
    y = 'FDR',
    legendPosition = "right",
    legendLabSize = 12,
    legendIconSize = 3,
    labSize = 3,
    subtitle = "POLA vs OPLA",
    pCutoff = 0.01, 
    FCcutoff = 0.2)
```

``` {R Table of differentially expressed genes, log2cpm}
# Decide tests go through the results of the differential gene expression analysis and catagorizes each gene into either significant up, down-regulation or no significnat change (1,-1 and 0 respectively). IMPORTANT - p.value and lfc are obviously the p-value and lfc cutoff for if genes are classed as differentially expressed. These can and probably should be changed and will have an impact on the results from here on. Currently set to same parameters as the volcano plots are.  

Decision_results1 <- decideTests(res1, method="global", adjust.method="BH", p.value=0.05, lfc = 0.2)
Decision_results2 <- decideTests(res2, method="global", adjust.method="BH", p.value=0.05, lfc = 0.2)
Decision_results3 <- decideTests(res3, method="global", adjust.method="BH", p.value=0.05, lfc = 0.2)

# I've combined the results of these three combine tests to have a column for each comparison 
Decision_results <- as_tibble(cbind(Decision_results1, Decision_results2, Decision_results3), rownames = "geneID")

# Here I've taken our average gene expression df we made and ask it to keep all the genes with a significant lfc as decided by the results of our decision tests
diffGenes.avg <- Avg.gene_expression[Decision_results[,2] !=0 | Decision_results[,3] !=0 | Decision_results[,4] !=0,]
diffGenes.avg.df <- as_tibble(diffGenes.avg)

# Created interactive and searchable table of all the significantly differentially expressed genes
datatable(diffGenes.avg.df, 
          extensions = c('KeyTable', "FixedHeader"), 
          caption = 'Table 2: log2 CPM of DEGs in Control, OPLA and POLA',
          options = list(keys = TRUE, searchHighlight = TRUE, pageLength = 10, lengthMenu = c("10", "25", "50", "100"))) %>%
  formatRound(columns=c(2:4), digits=2)

```

``` {R Graph Individual Genes}
# Pivot to tidy up data
diffGenes.avg.df.pivot <- pivot_longer(diffGenes.avg.df, 
                          cols = Control : POLA, 
                          names_to = "fat type", 
                          values_to = "log2(cpm)")

#  Filter for genes you want. grepl will pick out any with that phrase in it. Use " x|y|z|etc." notation to search for multiple genes. 
graph <- diffGenes.avg.df.pivot %>%
  dplyr::filter(grepl('AKR1D1|AKR1C4|SRD5A1|SRD5A2|11BHSD1', geneID))

# Graph error bars were a step too far for me as they got messed up with the facet wrap
ggbarplot(graph, "fat type", "log2(cpm)", add = "mean_se", palette = "npg", legend = "none", fill = "fat type", xlab = "Media", ylab = "counts per million (log2)") +
  facet_wrap(~geneID) +
  scale_y_continuous(expand = c(0,0), limits = c(0, 12)) +
theme(axis.text.x = element_text(color = "grey20", size = 12),
axis.text.y = element_text(color = "grey20", size = 12),
axis.title.x = element_text(color = "grey20", size = 16),
axis.title.y = element_text(color = "grey20", size = 16))
#ggsave('C:/Users/fwestcott/OneDrive - Nexus365/Desktop/RNAseq Graph2.png')

```

``` {R Venn Diagram}
#This is exactly the same as before I have just used the decision results from earlier to filter for up and down regulated genes instead of doing it manually

venn_opla_up <- toptable1[Decision_results[,2] >0,]
venn_opla_down <- toptable1[Decision_results[,2] <0,]
venn_pola_up <- toptable2[Decision_results[,3] >0,]
venn_pola_down <- toptable2[Decision_results[,3] <0,]

venn_up <-merge(venn_opla_up,venn_pola_up, by= "geneID", all.x=TRUE, all.y=TRUE)

OPLA_up <- venn_up[!is.na(venn_up$logFC.x), "geneID"]
POLA_up <- venn_up[!is.na(venn_up$logFC.y), "geneID"]
input_up <-list("Upregulated in OPLA"=OPLA_up, "Upregulated in POLA"=POLA_up)
venn(input_up)


venn_down <-merge(venn_opla_down,venn_pola_down, by= "geneID", all.x=TRUE, all.y=TRUE)

OPLA_down <- venn_down[!is.na(venn_down$logFC.x), "geneID"]
POLA_down <- venn_down[!is.na(venn_down$logFC.y), "geneID"]
input_down <-list("Downregulated in OPLA"=OPLA_down, "Downregulated in POLA"=POLA_down)
venn(input_down)

```


``` {R Clustering}
# I'm again using that big Decision results table to filter out all non-significant differential expressed genes from our log2cpm data
diffGenes <- log2.cpm.filtered.norm.df[Decision_results[,2] !=0 | Decision_results[,3] !=0 | Decision_results[,4] !=0,] %>% 
  column_to_rownames(var="geneID") %>%
  as.matrix

# This is just stuff for the dendrogram to do clustering. Trust me, I copied it off the internet.
clustRows <- hclust(as.dist(1-cor(t(diffGenes), method="pearson")), method="complete")
clustColumns <- hclust(as.dist(1-cor(diffGenes, method="spearman")), method="complete")

# Change K = it you want to cluster you genes into more than two (change it and you'll see what I mean)
module.assign <- cutree(clustRows, k=2)

# Boring code just to get the little colour bar by the side of the genes
module.color <- rainbow(length(unique(module.assign)), start=0.1, end=0.9) 
module.color <- module.color[as.vector(module.assign)]

# Actual heatmap
heatmap.2(diffGenes, 
          Rowv=as.dendrogram(clustRows), 
          Colv=as.dendrogram(clustColumns),
          RowSideColors=module.color,
          scale='row', labRow=NA,
          density.info="none", trace="none",  
          cexRow=1, cexCol=1, margins=c(8,20)) 

```

``` {R Heatmap of selected genes}
# Change whatever is in the filter function to change which genes are plotted
selected_genes <- diffGenes %>%
  as_tibble(rownames = "geneID") %>%
dplyr::filter(grepl('TRBC1|IL17RC|CDHR5|ALDOA|CLDN3|IGKV1OR2‐108|CHI3L1|EMILIN1|SCD|JUND|MYOT|BTRC|PDE4DIP|GART|RPGRIP1L|EEA1|GRIK2|TLX3|IGF2BP3|GTSE1',geneID))%>%
 column_to_rownames(var="geneID") %>%
  as.matrix

#TRBC1|IL17RC|CDHR5|ALDOA|CLDN3|IGKV1OR2‐108|CHI3L1|EMILIN1|SCD|JUND

#png(filename = "C:/Users/fwestcott/OneDrive - Nexus365/Desktop/Outputs/Transfer-report/figures/ATG heatmap.png", width = 800, height = 800, units = "px", pointsize = 17, bg = "white")

# I stripped this one down a lot because otherwise was a bit OTT
heatmap.2(selected_genes, 
          Rowv= NULL, 
          Colv=NULL,
          dendrogram = 'none',
          scale='row',
          density.info="none", trace="none",  
          cexRow= 1.4, cexCol=0.9, margins=c(1,20)
          )

#dev.off()

```

``` {R OPLA function enrichment analysis}

diffGenes.OPLA <- toptable1[Decision_results[,2] !=0,] %>%
rename("SYMBOL" = geneID)

entrez <- bitr(diffGenes.OPLA$SYMBOL, fromType = "SYMBOL",
                 toType = "ENTREZID",
                 OrgDb="org.Hs.eg.db")

deg_opla_enrich <- diffGenes.OPLA %>% left_join(entrez,.,by="SYMBOL")

opla_kegg_gene_list <- deg_opla_enrich$logFC

names(opla_kegg_gene_list) <- deg_opla_enrich$ENTREZID

opla_kegg_gene_list<-na.omit(opla_kegg_gene_list)

opla_kegg_gene_list = sort(opla_kegg_gene_list, decreasing = TRUE)

opla_kk <- gseKEGG(geneList     = opla_kegg_gene_list,
               organism     = "hsa")

dotplot(opla_kk, showCategory = 3, title = "OPLA Enriched Pathways vs Control" , split=".sign") + facet_grid(.~.sign)
  ggsave("C:/Users/fwestcott/OneDrive - Nexus365/Desktop/Outputs/Transfer-report/figures/OPLA GSEA.png")
```


``` {R OPLA pathway enrichment, warning=FALSE}
setwd("C:/Users/fwestcott/OneDrive - Nexus365/Desktop/Results/OPLA-POLA-huh7-RNAseq/Pathview")

#This is the same as before
fc.vector <- as.numeric(deg_opla_enrich[, 3])
names(fc.vector) <- (deg_opla_enrich[, 2])
de.names <- names(fc.vector)

pathview(gene.data  = fc.vector, pathway.id = "hsa04142",species  = "hsa")

```



``` {R POLA functional enrichment analysis}

diffGenes.POLA <- toptable2[Decision_results[,3] !=0,] %>%
rename("SYMBOL" = geneID)

entrez <- bitr(diffGenes.POLA$SYMBOL, fromType = "SYMBOL",
                 toType = "ENTREZID",
                 OrgDb="org.Hs.eg.db")

deg_pola_enrich <- diffGenes.POLA %>% left_join(entrez,.,by="SYMBOL")

pola_kegg_gene_list <- deg_pola_enrich$logFC

names(pola_kegg_gene_list) <- deg_pola_enrich$ENTREZID

pola_kegg_gene_list<-na.omit(pola_kegg_gene_list)

pola_kegg_gene_list = sort(pola_kegg_gene_list, decreasing = TRUE)

pola_kk <- gseKEGG(geneList     = pola_kegg_gene_list,
               organism     = "hsa")

dotplot(pola_kk, showCategory = 3, title = "POLA Enriched Pathways vs Control" , split=".sign")  + facet_grid(.~.sign)
ggsave("C:/Users/fwestcott/OneDrive - Nexus365/Desktop/Outputs/Transfer-report/figures/POLA GSEA.png")
```

``` {R POLA pathway enrichment, warning=FALSE}
setwd("C:/Users/fwestcott/OneDrive - Nexus365/Desktop/Results/OPLA-POLA-huh7-RNAseq/Pathview")

#This is the same as before
fc.vector <- as.numeric(deg_pola_enrich[, 3])
names(fc.vector) <- (deg_pola_enrich[, 2])
de.names <- names(fc.vector)

pathview(gene.data  = fc.vector, pathway.id = "hsa04144",species  = "hsa")

```

``` {R POLA vs OPLA function enrichment analysis}

diffGenes.POLAOPLA <- toptable3[Decision_results[,4] !=0,] %>%
rename("SYMBOL" = geneID)

entrez <- bitr(diffGenes.POLAOPLA$SYMBOL, fromType = "SYMBOL",
                 toType = "ENTREZID",
                 OrgDb="org.Hs.eg.db")

deg_pola_opla_enrich <- diffGenes.POLAOPLA %>% left_join(entrez,.,by="SYMBOL")

setwd("C:/Users/fwestcott/OneDrive - Nexus365/Desktop/Results/OPLA-POLA-huh7-RNAseq/Pathview")

#This is the same as before
fc.vector <- as.numeric(deg_pola_opla_enrich[, 3])
names(fc.vector) <- (deg_pola_opla_enrich[, 2])
de.names <- names(fc.vector)

pathview(gene.data  = fc.vector, pathway.id = "hsa04144",species  = "hsa")


```
``` {R Extra analysis}

#Genes significantly differentially expressed between OPLA and POLA but both in the same way compared to controls

Decision_results_extra <- Decision_results
Decision_results_extra$Extra <- if_else(Decision_results_extra$`-1*OPLA 1*POLA` == 0, 0, 
                                        if_else(Decision_results_extra$`-1*Control 1*OPLA` == Decision_results_extra$`-1*Control 1*POLA`, 1, 0))

diffGenes.Extra <- toptable3[Decision_results_extra$Extra !=0,] %>%
  rename("SYMBOL" = geneID)

entrez <- bitr(diffGenes.Extra$SYMBOL, fromType = "SYMBOL",
                 toType = "ENTREZID",
                 OrgDb="org.Hs.eg.db")

deg_extra_enrich <- diffGenes.Extra %>% left_join(entrez,.,by="SYMBOL")

setwd("C:/Users/fwestcott/OneDrive - Nexus365/Desktop/Results/OPLA-POLA-huh7-RNAseq/Pathview")

#This is the same as before
fc.vector <- as.numeric(deg_extra_enrich[, 3])
names(fc.vector) <- (deg_extra_enrich[, 2])
de.names <- names(fc.vector)

pathview(gene.data  = fc.vector, pathway.id = "hsa04140",species  = "hsa")

```